README.md
Cryptotronix Hashlet
=====

[![Build Status](https://travis-ci.org/cryptotronix/hashlet.png)](https://travis-ci.org/cryptotronix/hashlet)
<a href="https://scan.coverity.com/projects/1908">
  <img alt="Coverity Scan Build Status"
       src="https://scan.coverity.com/projects/1908/badge.svg"/>
</a>

Status
---

As of 7 February, 2014 the 1.0.0 release is considered a second Beta release and is available on the [release](http://download.savannah.gnu.org/releases/hashlet/) page.  It has been tested on a BeagleBone Black, rev 5AC running Debian Wheezy and a Raspberry Pi running Raspian.  It *should* work with Angstrom, but has not been tested.  Beta testers are welcome!

Building
----

This project uses Autotools and you need that installed to configure and build the executable.  I am mainly developing on a BeagleBone Black using Debian.

If you pull this repo (i.e. a non-release), you will need the following dependencies:
- autotools (i.e. automake, autconf, and libtool)
- Flex and Bison
- texinfo (for the documentation if you so desire)

The run time dependencies are:
- libgcrypt

Hardware
---

The hardware is available from [Cryptotronix](http://cryptotronix.com/products/hashlet/).  We are an open hardware company, so see the `hardware` folder for the design to make this yourself.

BOM
---

The chip on the Hashlet is Atmel's ATSHA204.

Running
---

see `./hashlet --help` for full details.  The default I2C bus is `/dev/i2c-1` and this can be changed with the `-b` option.  On some BBB, the bus is `/dev/i2c-2`.  See this [blog post](http://datko.net/2013/11/03/bbb_i2c/) for further details on BBB I2C.

Root
---

You'll need to run as root to access `/dev/i2c*` initially.  You can change this by adding your user to the `i2c` group with:

`sudo usermod -aG i2c user`

Or:

`sudo chmod o+rw /dev/i2c*`


Currently supported commands:

### state
```bash
./hashlet state
Factory
```

This is the first command you should run and verify it's in the Factory state.  This provides the assurance that the device has not been tampered during transit.

### personalize
```bash
./hashlet personalize
```

With the key import feature:

```bash
./hashlet personalize -f keys.txt
```

This is the second command you should run.  On success it will not output anything.  Random keys are loaded into the device and saved to `~/.hashlet` as a backup.  Don't lose that file.  Keys from another hashlet can be imported with the `-f` option, where the file is not also named `~/.hashlet`.

### random
```bash
./hashlet random
62F95589AC76855A8F9204C9C6B8B85F06E6477D17C3888266AEE8E1CBD65319
```

This command also takes the `-B` parameter which allows you to specify the number of bytes you want back from the random number generator.

### mac
```bash
./hashlet mac --file test.txt
mac       : C3466ABB8640B50938B260E17D86489D0EBB3F9C8009024683CB225FFFD3B4E4
challenge : 9F0751C90770E6B40E34BA8E06EFE453FAA46B5FB26925FFBD664FAF951D000A
meta      : 08000000000000000000000000
```

On success it will output three parameters:

1. mac: (aka challenge response) The result of the operation
2. challenge: This is the input to the Hashlet, after a SHA256 digest
3. meta: Meta data that must accompany the result

### check-mac
```bash
./hashlet check-mac -r C3466ABB8640B50938B260E17D86489D0EBB3F9C8009024683CB225FFFD3B4E4 -c 9F0751C90770E6B40E34BA8E06EFE453FAA46B5FB26925FFBD664FAF951D000A -m 08000000000000000000000000
```

Checks the MAC that was produced by the Hashlet.  On success, it will with an exit value of 0.

### offline-verify
```bash
./hashlet offline-verify -c 322B3FFC3BE16B4CC5B445F8E666D0BA5C5E676D00FABD2308AD51243FA0B067 -r FB19B1C63161B6C34CA9D291D1CD16F98247BBA9A298775F795161BEB95BB6EF
```

On success, it will output an exit code of 0, otherwise it will fail.  The point of this command is that a remote server can verify the MAC from the Hashlet without a device.  The keys are written to `~/.hashlet` upon personalization and if this file is store on the server, it can verify a MAC.

The workflow goes like this:

1. Mac some data to produce a challenge response.
2. Send the challenge and MAC to the remote server, which has the key store file.
3. Perform offline-verify on the remote server.

### hmac
```bash
hashlet hmac -f ChangeLog
CD0765AB1F94698E0EACDE22C10F362E925F4F4017B75FDE5AB3124FCEBE9754
```
```bash
echo test | ./hashlet hmac
9CAF53A19F4A8D751F9D03ED3991EF648DC9246D2B329D9A650307212D994326
```

Performs HMAC-256. Your input data is combined with data on the device, so this isn't a pure HMAC-256. Your data will be applied to SHA-256 and then combined, per the datasheet, with device data. That result is fed into HMAC-256.

### offline-hmac
```bash
hashlet offline-hmac -r CD0765AB1F94698E0EACDE22C10F362E925F4F4017B75FDE5AB3124FCEBE9754 -f ChangeLog
echo $?
0
```
```bash
echo test | ./hashlet offline-hmac -r 9CAF53A19F4A8D751F9D03ED3991EF648DC9246D2B329D9A650307212D994326
echo $?
0
```

Similar to `offline-verify`, verifies the hmac operation using the key file from `~/.hashlet`.
### serial-num
```bash
./hashlet serial-num
0123XXXXXXXXXXXXEE
```
X's indicate the unique serial number.

Options
---

Options are listed in the `--help` command, but a useful one, if there are issues, is the `-v` option.  This will dump all the data that
travels across the I2C bus with the device.


Design
---

In the `hardware` folder, one should find the design files for the Hashlet.  The IC on the hashlet is the [Atmel ATSHA204](http://www.atmel.com/Images/Atmel-8740-CryptoAuth-ATSHA204-Datasheet.pdf).

Support
---

IRC: Join the `#cryptotronix` channel on freenode.

Mailing lists: `hashlet-announce` and `hashlet-users` are open for subscriptions [here](https://savannah.nongnu.org/mail/?group=hashlet).

Contributing
---
See the wiki page on [contributing](https://github.com/cryptotronix/hashlet/wiki/Contributing).


NEWS
Hashlet NEWS -- history of user-visible changes

Copyright (C) 2014 Cryptotronix, LLC.

* Hashlet 1.1.0 - Release - April 2014

This is the fist non-beta release.

** New features

- random now takes as an option, -B, which specifies the number of
  bytes to return

- Added hmac and offline-hmac to produce a HMAC-256 on the device and
  to verify it without the device.

* Hashlet 1.0.0 - Beta Release - February 2014

This is the second Beta release of this software.

** Major Changes from Previous version

- CLI changed to default to /dev/i2c-1.  Other interfaces can be
  specified with the -b option.

** New Features

- Adds import key feature: personalize -f file.txt, which will allow
  cloning of keys from one device to the next.

- Changes the key slot configuration such that 2 keys can be updated
  with the new write command, with valid mac (currently not working).

- 2 Keys are designated as read / write for user data or storage.

- Adds write command to write 32 bytes to supported key slots.

* Hashlet 0.1.0 - Beta Release - December 2013

The initial release of Hashlet is concerned a Beta release.

** Features

*** Command line interface

The following commands are supported:

**** personalize
**** random
**** mac
**** check-mac
**** offline-verify
**** serial-num
**** state

*** Sets Defaults Key Usage

See the documentation for further details.

--------------------------------------------------------------------
This file is part of Hashlet.

Hashlet is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

Hashlet is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Hashlet.  If not, see <http://www.gnu.org/licenses/>.

Local variables:
mode: outline
paragraph-separate: "[ 	]*$"
end:


src/main.c
/*
 * Copyright (C) 2013 Cryptotronix, LLC.
 *
 * This file is part of Hashlet.
 *
 * Hashlet is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * Hashlet is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Hashlet.  If not, see <http://www.gnu.org/licenses/>.
 *
 */


#include <stdlib.h>
#include <argp.h>
#include <stdbool.h>
#include <assert.h>
#include <string.h>
#include "config.h"
#include "hashlet.h"


const char *argp_program_version = PACKAGE_VERSION;

const char *argp_program_bug_address = PACKAGE_BUGREPORT;

/* Program documentation. */
static char doc[] =
  "Hashlet is a program to interface to the Cryptotronix Hashlet.\n\n"
  "Currently implemented Commands:\n\n"
  "random        --  Retrieves 32 bytes of random data from the device.\n"
  "serial-num    --  Retrieves the device's serial number.\n"
  "mac           --  Calculates a SHA-256 digest of your input data and then\n"
  "                  sends that digest to the device to be mac'ed with a key\n"
  "                  other internal data\n"
  "slot-config   --  Prints out the slot configuration for a given key.\n"
  "                  Use with key-slot option\n";

/* A description of the arguments we accept. */
static char args_doc[] = "i2c_bus command";

#define OPT_UPDATE_SEED 300


/* The options we understand. */
static struct argp_option options[] = {
  {"verbose",  'v', 0,      0,  "Produce verbose output" },
  {"quiet",    'q', 0,      0,  "Don't produce any output" },
  {"silent",   's', 0,      OPTION_ALIAS },
  {"update-seed", OPT_UPDATE_SEED, 0, 0,
     "Updates the random seed.  Only applicable to certain commands"},
  {"key-slot", 'k', "SLOT",      0,  "The internal key slot to use."},
  {"output",   'o', "FILE", 0,
   "Output to FILE instead of standard output" },
  { 0 }
};


#define NUM_ARGS 2

/* Used by main to communicate with parse_opt. */
struct arguments
{
  char *args[NUM_ARGS];
  int silent, verbose;
  bool update_seed;
  char *output_file;
  unsigned int key_slot;
  struct mac_mode_encoding mac_mode;
};

/* Parse a single option. */
static error_t
parse_opt (int key, char *arg, struct argp_state *state)
{
  /* Get the input argument from argp_parse, which we
     know is a pointer to our arguments structure. */
  struct arguments *arguments = state->input;
  int slot;

  switch (key)
    {
    case 'q': case 's':
      arguments->silent = 1;
      break;
    case 'v':
      arguments->verbose = 1;
      set_log_level(DEBUG);
      break;
    case 'o':
      arguments->output_file = arg;
      break;
    case OPT_UPDATE_SEED:
      arguments->update_seed = true;
      break;
    case 'k':
      slot = atoi(arg);
      if (slot < 0 || slot > 15)
        argp_usage (state);

      arguments->key_slot = slot;
      break;
    case ARGP_KEY_ARG:
      if (state->arg_num > NUM_ARGS)
        /* Too many arguments. */
        argp_usage (state);

      arguments->args[state->arg_num] = arg;

      break;

    case ARGP_KEY_END:
      if (state->arg_num < NUM_ARGS)
        /* Not enough arguments. */
        argp_usage (state);
      break;

    default:
      return ARGP_ERR_UNKNOWN;
    }
  return 0;
}

/* Our argp parser. */
static struct argp argp = { options, parse_opt, args_doc, doc };

/* Default ADDR of Hashlet */

const int ADDR = 0b1100100;


void output_hex(FILE *stream, struct octet_buffer buf)
{
  assert(NULL != buf.ptr);
  assert(NULL != stream);

  unsigned int i = 0;

  for (i = 0; i < buf.len; i++)
    {
      fprintf(stream, "%02X", buf.ptr[i]);
    }

  fprintf(stream, "\n");

}

int
main (int argc, char **argv)
{
  struct arguments arguments;
  struct octet_buffer response;

  /* Default values. */
  arguments.silent = 0;
  arguments.verbose = 0;
  arguments.output_file = "-";
  arguments.update_seed = false;
  arguments.key_slot = 0;

  /* Default MAC mode */
  arguments.mac_mode.use_serial_num = false;
  arguments.mac_mode.use_otp_0_7 = false;
  arguments.mac_mode.use_otp_0_10 = false;
  arguments.mac_mode.temp_key_source_flag = false;
  arguments.mac_mode.use_first_32_temp_key = false;
  arguments.mac_mode.use_second_32_temp_key = false;

  /* Parse our arguments; every option seen by parse_opt will
     be reflected in arguments. */
  argp_parse (&argp, argc, argv, 0, 0, &arguments);

  int fd;

  fd = hashlet_setup(arguments.args[0], ADDR);

  if (fd < 0)
    exit(fd);

  if (0 == strcmp(arguments.args[1], "random"))
    {

      response = get_random(fd, arguments.update_seed);
      print_hex_string("Random:", response.ptr, response.len);
      output_hex(stdout, response);
      free_octet_buffer(response);

    }
  else if(0 == strcmp(arguments.args[1], "serial-num"))
  {
      response = get_serial_num(fd);
      output_hex(stdout, response);
      free_octet_buffer(response);
  }
  else if(0 == strcmp(arguments.args[1], "mac"))
    {
      uint8_t canned[32];
      memset(canned, 0xFF, 32);
      struct octet_buffer challenge = {canned, 32};

      response = perform_mac(fd, arguments.mac_mode,
                             arguments.key_slot, challenge);
      output_hex(stdout, response);
      free_octet_buffer(response);
    }
  else if(0 == strcmp(arguments.args[1], "slot-config"))
    {
      printf ("TODO\n");

    }
  else
    {
      printf("Invalid command, exiting.  Try --help\n");
      hashlet_teardown(fd);
      exit(1);
    }

  /* printf ("ARG1 = %s\nARG2 = %s\nOUTPUT_FILE = %s\n" */
  /*         "VERBOSE = %s\nSILENT = %s\n", */
  /*         arguments.args[0], arguments.args[1], */
  /*         arguments.output_file, */
  /*         arguments.verbose ? "yes" : "no", */
  /*         arguments.silent ? "yes" : "no"); */

  hashlet_teardown(fd);
  exit (0);
}
